### ISSUE TITLE
Run: UNABLE TO RETRIEVE SPWN_REQ FOR JOB

### ISSUE THREAD

**Issue by tonycurtis** [openpmix/prrte#967]
Thank you for taking the time to submit an issue!

## Background information

With latest HEAD, seeing  message as described below

### What version of the PMIx Reference Server are you using? (e.g., v1.0, v2.1, git master @ hash, etc.)

git master @ b22728da461141aa00ebd91f705eb262e1e081d7

##### What version of PMIx are you using? (e.g., v1.2.5, v2.0.3, v2.1.0, git branch name and hash, etc.)

3.2.3 (and git HEAD master @ 74327dd61ffa01d7e0a731887118c28aefd78683)

### Please describe the system on which you are running

* Operating system/version: CEntOS 8
* Computer hardware: aarch64 (a64fx)
* Network type: IB

-----------------------------

## Details of the problem

Before and after regular output...about a 10 second delay before anything happens.

```
[login1:218682] PRTE ERROR: Not found in file ../../prrte-git/src/prted/pmix/pmix_server_dyn.c at line 72
[login1:218682] UNABLE TO RETRIEVE SPWN_REQ FOR JOB prterun-login1-218682@1 [room=4095]
...
[login1:218682] PRTE ERROR: Not found in file ../../prrte-git/src/prted/pmix/pmix_server_dyn.c at line 72
[login1:218682] UNABLE TO RETRIEVE SPWN_REQ FOR JOB prterun-login1-218682@1 [room=4095]
```

**Comment by rhc54** [comment]
Could you tell me what you did to get that error? Is it with one node, multiple nodes? What was the prun cmd line? Etc.

**Comment by tonycurtis** [comment]
e.g.

```
salloc -p ... -N 2 --ntasks-per-node=1 prterun --personality shmem ./a.out
```

**Comment by tonycurtis** [comment]
Also happening about 50% of the time using `prterun` to launch Open-MPI programs

**Comment by rhc54** [comment]
Odd - so the job runs, but start is delayed? And then you get the second output after the job ends, following another delay?

**Comment by tonycurtis** [comment]
Just the initial ~10 second delay with no output, then everything appears

**Comment by rhc54** [comment]
Can you try it with the following addition to the `prterun` cmd line: `--prtemca plm ssh`? I'm not able to replicate locally, but I'm not using Slurm.

**Comment by tonycurtis** [comment]
```
$ time salloc -p short -n 2 --ntasks-per-node=1 prterun --prtemca plm ssh  ./a.out
salloc: Granted job allocation 16769
fj169: 0 of 1 on core #0
fj168: 0 of 1 on core #0
salloc: Relinquishing job allocation 16769

real	0m12.642s
user	0m0.041s
sys	0m0.098s
```

So it runs, but with the ~10 sec delay, and unconnected MPI processes

**Comment by rhc54** [comment]
Is the 10sec happening after you are granted the allocation? I'm wondering if you are seeing some artifact of the `salloc`. Maybe put the `time` cmd before the `prterun`?

**Comment by tonycurtis** [comment]
Still about ~10-12 seconds.  With OMPI 4.1.1, start-up is < 1 sec.

**Comment by rhc54** [comment]
Huh - strange. But you don't get the error message if you use `ssh`?

**Comment by tonycurtis** [comment]
No error message, but both ranks report "0 of 1" so I guess they think they are separate programs.

**Comment by rhc54** [comment]
Yeah - they think they are singletons. Weird. I'll have to dig a bit and see if I can figure out a way to replicate it. Things are working fine over here and on the CI, so this may take a bit to figure out.

**Comment by tonycurtis** [comment]
Latest OMPI HEAD not showing this behavior BTW

**Comment by rhc54** [comment]
Looking at your cmd line - is it possible you are launching _two_ copies of `prterun`, one on each node? IIUC, that is what your cmd line is telling Slurm to do. I can believe that would indeed cause major problems.

**Comment by tonycurtis** [comment]
salloc + mpi{run,exec} works fine, shouldn't prterun work in the same way?  salloc starts the launch on the same node (login1).  Note the SPWN messages referencing login1.

**Comment by tonycurtis** [comment]
Here's a couple of as-simple-as-possible examples.  Note the 5 second or so delay with `prterun` on 1 node.  If I try with 2 nodes, 1 rank per node, the delay is again about 10 seconds.  OMPI 4.1.1 again.

```
login1[2](~) salloc -p short -n 1 time prterun hostname
salloc: Granted job allocation 16794
fj023
0.05user 0.07system 0:05.46elapsed 2%CPU (0avgtext+0avgdata 15680maxresident)k
0inputs+0outputs (0major+374minor)pagefaults 0swaps
salloc: Relinquishing job allocation 16794

login1[2](~) salloc -p short -n 1 time mpirun hostname
salloc: Granted job allocation 16795
fj166
0.04user 0.10system 0:00.63elapsed 23%CPU (0avgtext+0avgdata 23360maxresident)k
0inputs+0outputs (0major+680minor)pagefaults 0swaps
salloc: Relinquishing job allocation 16795
```

and then later

```
login1[2](~) salloc -p short -N 2 --ntasks-per-node=1  time prterun hostname
salloc: Granted job allocation 16796
fj021
[login1:138149] PRTE ERROR: Not found in file ../../prrte-git/src/prted/pmix/pmix_server_dyn.c at line 72
[login1:138149] UNABLE TO RETRIEVE SPWN_REQ FOR JOB prterun-login1-138149@1 [room=4095]
fj022
[login1:138149] PRTE ERROR: Not found in file ../../prrte-git/src/prted/pmix/pmix_server_dyn.c at line 72
[login1:138149] UNABLE TO RETRIEVE SPWN_REQ FOR JOB prterun-login1-138149@1 [room=4095]
0.03user 0.09system 0:10.47elapsed 1%CPU (0avgtext+0avgdata 15680maxresident)k
0inputs+0outputs (0major+376minor)pagefaults 0swaps
salloc: Relinquishing job allocation 16796
```

**Comment by rhc54** [comment]
Dunno - I never trust doing salloc and a cmd on the same line. I always get the allocation and then run a cmd separately. However, I try not to become a Slurm expert, so if you say that should work, I'll take it from there. I'm mostly looking for differences as I still can't get this to replicate anywhere.

**Comment by rhc54** [comment]
This is what I get when running across three nodes:
```
$ time prterun --personality shmem ./hello
Client ns prterun-rhc-node01-20430@1 rank 3 pid 20439: Running on host rhc-node01 localrank 3
prterun-rhc-node01-20430@1:3 - local peers 0,1,2,3,4,5,6,7,8,9
Client ns prterun-rhc-node01-20430@1 rank 6 pid 20442: Running on host rhc-node01 localrank 6
prterun-rhc-node01-20430@1:6 - local peers 0,1,2,3,4,5,6,7,8,9
Client ns prterun-rhc-node01-20430@1 rank 5 pid 20441: Running on host rhc-node01 localrank 5
prterun-rhc-node01-20430@1:5 - local peers 0,1,2,3,4,5,6,7,8,9
Client ns prterun-rhc-node01-20430@1 rank 4 pid 20440: Running on host rhc-node01 localrank 4
prterun-rhc-node01-20430@1:4 - local peers 0,1,2,3,4,5,6,7,8,9
Client ns prterun-rhc-node01-20430@1 rank 7 pid 20443: Running on host rhc-node01 localrank 7
prterun-rhc-node01-20430@1:7 - local peers 0,1,2,3,4,5,6,7,8,9
Client ns prterun-rhc-node01-20430@1 rank 2 pid 20438: Running on host rhc-node01 localrank 2
prterun-rhc-node01-20430@1:2 - local peers 0,1,2,3,4,5,6,7,8,9
Client ns prterun-rhc-node01-20430@1 rank 0 pid 20436: Running on host rhc-node01 localrank 0
prterun-rhc-node01-20430@1:0 - local peers 0,1,2,3,4,5,6,7,8,9
Client ns prterun-rhc-node01-20430@1 rank 8 pid 20444: Running on host rhc-node01 localrank 8
prterun-rhc-node01-20430@1:8 - local peers 0,1,2,3,4,5,6,7,8,9
Client ns prterun-rhc-node01-20430@1 rank 9 pid 20445: Running on host rhc-node01 localrank 9
prterun-rhc-node01-20430@1:9 - local peers 0,1,2,3,4,5,6,7,8,9
Client ns prterun-rhc-node01-20430@1 rank 1 pid 20437: Running on host rhc-node01 localrank 1
prterun-rhc-node01-20430@1:1 - local peers 0,1,2,3,4,5,6,7,8,9
Client ns prterun-rhc-node01-20430@1 rank 26 pid 926: Running on host rhc-node03 localrank 6
prterun-rhc-node01-20430@1:26 - local peers 20,21,22,23,24,25,26,27,28,29
Client ns prterun-rhc-node01-20430@1 rank 25 pid 929: Running on host rhc-node03 localrank 5
prterun-rhc-node01-20430@1:25 - local peers 20,21,22,23,24,25,26,27,28,29
Client ns prterun-rhc-node01-20430@1 rank 14 pid 940: Running on host rhc-node02 localrank 4
prterun-rhc-node01-20430@1:14 - local peers 10,11,12,13,14,15,16,17,18,19
Client ns prterun-rhc-node01-20430@1 rank 10 pid 936: Running on host rhc-node02 localrank 0
prterun-rhc-node01-20430@1:10 - local peers 10,11,12,13,14,15,16,17,18,19
Client ns prterun-rhc-node01-20430@1 rank 24 pid 927: Running on host rhc-node03 localrank 4
prterun-rhc-node01-20430@1:24 - local peers 20,21,22,23,24,25,26,27,28,29
Client ns prterun-rhc-node01-20430@1 rank 29 pid 931: Running on host rhc-node03 localrank 9
prterun-rhc-node01-20430@1:29 - local peers 20,21,22,23,24,25,26,27,28,29
Client ns prterun-rhc-node01-20430@1 rank 16 pid 942: Running on host rhc-node02 localrank 6
prterun-rhc-node01-20430@1:16 - local peers 10,11,12,13,14,15,16,17,18,19
Client ns prterun-rhc-node01-20430@1 rank 15 pid 941: Running on host rhc-node02 localrank 5
prterun-rhc-node01-20430@1:15 - local peers 10,11,12,13,14,15,16,17,18,19
Client ns prterun-rhc-node01-20430@1 rank 20 pid 922: Running on host rhc-node03 localrank 0
prterun-rhc-node01-20430@1:20 - local peers 20,21,22,23,24,25,26,27,28,29
Client ns prterun-rhc-node01-20430@1 rank 18 pid 943: Running on host rhc-node02 localrank 8
prterun-rhc-node01-20430@1:18 - local peers 10,11,12,13,14,15,16,17,18,19
Client ns prterun-rhc-node01-20430@1 rank 27 pid 928: Running on host rhc-node03 localrank 7
prterun-rhc-node01-20430@1:27 - local peers 20,21,22,23,24,25,26,27,28,29
Client ns prterun-rhc-node01-20430@1 rank 12 pid 938: Running on host rhc-node02 localrank 2
prterun-rhc-node01-20430@1:12 - local peers 10,11,12,13,14,15,16,17,18,19
Client ns prterun-rhc-node01-20430@1 rank 13 pid 939: Running on host rhc-node02 localrank 3
prterun-rhc-node01-20430@1:13 - local peers 10,11,12,13,14,15,16,17,18,19
Client ns prterun-rhc-node01-20430@1 rank 17 pid 944: Running on host rhc-node02 localrank 7
prterun-rhc-node01-20430@1:17 - local peers 10,11,12,13,14,15,16,17,18,19
Client ns prterun-rhc-node01-20430@1 rank 23 pid 925: Running on host rhc-node03 localrank 3
prterun-rhc-node01-20430@1:23 - local peers 20,21,22,23,24,25,26,27,28,29
Client ns prterun-rhc-node01-20430@1 rank 22 pid 923: Running on host rhc-node03 localrank 2
prterun-rhc-node01-20430@1:22 - local peers 20,21,22,23,24,25,26,27,28,29
Client ns prterun-rhc-node01-20430@1 rank 19 pid 945: Running on host rhc-node02 localrank 9
prterun-rhc-node01-20430@1:19 - local peers 10,11,12,13,14,15,16,17,18,19
Client ns prterun-rhc-node01-20430@1 rank 28 pid 930: Running on host rhc-node03 localrank 8
prterun-rhc-node01-20430@1:28 - local peers 20,21,22,23,24,25,26,27,28,29
Client ns prterun-rhc-node01-20430@1 rank 11 pid 937: Running on host rhc-node02 localrank 1
prterun-rhc-node01-20430@1:11 - local peers 10,11,12,13,14,15,16,17,18,19
Client ns prterun-rhc-node01-20430@1 rank 21 pid 924: Running on host rhc-node03 localrank 1
prterun-rhc-node01-20430@1:21 - local peers 20,21,22,23,24,25,26,27,28,29

real	0m1.301s
user	0m0.057s
sys	0m0.204s
```
Fully repeatable to within 0.05 seconds of real time.

**Comment by tonycurtis** [comment]
Hmmm.  salloc + run-command has worked everywhere I have tried it.  Our setup doesn't do srun + --mpi=pmix for some reason, but salloc + mpirun (or my OpenSHMEM equiv which drops to OMPI mpirun) has always been good.  So, as noted before, the salloc + launch is running the program from a login/"other" node that is not part of the job allocation.

**Comment by rhc54** [comment]
Hmmm...ok, maybe that's the diff. I can try to replicate, though it may take a bit to try and recreate that setup. Probably not until tomorrow or Fri.

**Comment by tonycurtis** [comment]
No hurry, OMPI 4.1.1 is my go-to launcher, just testing the latest pmix/prte combo

**Comment by rhc54** [comment]
Dang - running with `prterun` on a "login" node and two MPI procs, each running on a separate compute node:
```
[mpiuser@rhc-node01 test]$ time prterun --hostfile ../hosts.txt --map-by ppr:1:node ./hello
Client ns prterun-rhc-node01-75131@1 rank 1 pid 689: Running on host rhc-node03 localrank 0
prterun-rhc-node01-75131@1:1 - local peers 1
Client ns prterun-rhc-node01-75131@1 rank 0 pid 709: Running on host rhc-node02 localrank 0
prterun-rhc-node01-75131@1:0 - local peers 0

real	0m0.892s
user	0m0.017s
sys	0m0.033s
$
```
I've run this in a loop for 100 times without much variation in behavior. So I'm at a bit of a loss. Only thing I can figure is that there is some behavior associated with running under Slurm, but I have no way of tracking that down.

@abouteiller I believe you indicated you see similar behavior somewhere - can you please help track this down?

Note: I do see one oddity - the rank=0 process reports no local peers (which is correct), while rank=1 sees one. I'll have to look into that as a separate problem.

**Comment by rhc54** [comment]
> Note: I do see one oddity - the rank=0 process reports no local peers (which is correct), while rank=1 sees one. I'll have to look into that as a separate problem.

User error on my part - it is reporting the identity of the local peer, not the number of them.

**Comment by rhc54** [comment]
@tonycurtis When you get a chance, let's try a couple of things to help debug this:

* add "--prtemca prte_do_not_resolve 1" to your prterun cmd line
* if that doesn't solve the problem, then let's try adding "--prtemca state_base_verbose 5" and post the output. This will tell us the state in which the time is being consumed.

**Comment by tonycurtis** [comment]
The do_not_resolve seems to make things work.

**Comment by rhc54** [comment]
Hooray! Problem is a DNS one that we've seen before - historically, we do not resolve nodes on managed allocations because it can take too long. I've restored that behavior in the referenced PR.

Thanks for the patience and the assist!

**Comment by tonycurtis** [comment]
Actual DNS lookups or gethostby*() calls?

**Comment by rhc54** [comment]
It's a call to `getaddrinfo` - we use it to identify an actual local interface when multiple string names are used for the localhost. Unfortunately, since we don't know which ones _might_ be local versus those that definitely are _not_ local, we wind up running it on every node in the allocation. ☹️

**Comment by tonycurtis** [comment]
What's the equiv env var for the do_not_resolve setting?

Also BTW the description for this in `src/runtime/prte_mca_params.c` looks like part of the text missing (default value).  And there's some variation in use of "[]" vs. "()" for the default setting description in this file.

**Comment by rhc54** [comment]
`PRTE_MCA_prte_do_not_resolve`

I'll look into the params file.

**Comment by tonycurtis** [comment]
Thanks (went through a few naming attempts)

**Comment by rhc54** [comment]
FWIW: you cannot set it for managed allocations - it is hard-coded to "true".
